lib/composite_primary_keys/composite_predicates.rb

```diff
@@ -51,9 +51,58 @@ def cpk_join_predicate(table1, key1, table2, key2)
     end
 
     def cpk_in_predicate(table, primary_keys, ids)
+      if primary_keys.length == 2
+        cpk_in_predicate_with_grouped_keys(table, primary_keys, ids)
+      else
+        cpk_in_predicate_with_non_grouped_keys(table, primary_keys, ids)
+      end
+    end
+
+    def cpk_in_predicate_with_non_grouped_keys(table, primary_keys, ids)
       and_predicates = ids.map do |id|
         cpk_id_predicate(table, primary_keys, id)
       end
+
+      cpk_or_predicate(and_predicates)
+    end
+
+    def cpk_in_predicate_with_grouped_keys(table, primary_keys, ids)
+      keys_by_first_column_name = Hash.new { |hash, key| hash[key] = [] }
+      keys_by_second_column_name = Hash.new { |hash, key| hash[key] = [] }
+
+      ids.map.each do |first_key_part, second_key_part|
+        keys_by_first_column_name[first_key_part] << second_key_part
+        keys_by_second_column_name[second_key_part] << first_key_part
+      end
+
+      low_cardinality_column_name, high_cardinality_column_name, groups = \
+        if keys_by_first_column_name.size <= keys_by_second_column_name.size
+          [primary_keys.first, primary_keys.second, keys_by_first_column_name]
+        else
+          [primary_keys.second, primary_keys.first, keys_by_second_column_name]
+        end
+
+      and_predicates = groups.map do |low_cardinality_value, high_cardinality_values|
+        in_clause = table[high_cardinality_column_name].in(high_cardinality_values.compact)
```

Comment:
postgresql, mysql, oracle, etc do not allow null values in primary keys (in any field). So the compact call should be unnecessary as well as the include?(nil) check.

Reply:
I'm happy to remove it. But I think cpk can be used in situations where the columns defined in the ActiveRecord primary key are not the primary key in the database.

We do this in some situations (e.g. views), but we always require both parts to be present. I'm happy to remove this, if that is desirable.

Reply:
What happens in the current implementation with nil values? If they aren't supported now I would not add support for them now.

Reply:
The existing implementation of cpk_in_predicate (or in this PR cpk_in_predicate_with_non_grouped_keys) would create a queries like:

```sql
# pk [1, 1]
SELECT * FROM table WHERE pk1 = 1 AND pk2 = 1

# pk [1, nil]
SELECT * FROM table WHERE pk1 = 1 AND pk2 IS NULL

# pk [nil, 1]
SELECT * FROM table WHERE pk1 IS NULL AND pk2 = 1

# pk [nil, nil]
SELECT * FROM table WHERE pk1 IS NULL AND pk2 IS NULL
```

Arel transparently converts something like `table[column_name].eq(value)` to an `IS NULL` if the value is `nil`.

Reply:
Ah - then let's keep the null handling then.

Reply:
Ah - then let's keep the null handling then.

Reply:
However, the code checks the array twice for nil, once with `compact` and once with `'include?`. Let's just check once.

```
compacted = high_cardinality_values.compact
...
if high_cardinality_values.size > compacted.size
...
```

Reply:
Done, feel free to take a look.